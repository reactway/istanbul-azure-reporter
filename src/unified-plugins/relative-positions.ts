import { Attacher, Transformer } from "unified";
import { Position, Point, Parent } from "unist";

import { matchAll, NEWLINE_PATTERN } from "../helpers";
import { isNodeElement, isNodeText, modifyTreeFromDown, isNodeRoot } from "./helpers";
import { TextElement } from "./contracts";

export function getEndPoint(node: TextElement, position: Position): Point {
    const text = node.value;
    const matches = matchAll(NEWLINE_PATTERN, text);
    if (matches.length === 0) {
        return {
            line: position.start.line,
            column: position.start.column + text.length
        };
    }

    const lastMatch: RegExpMatchArray = matches[matches.length - 1];
    const lastLineOffset = lastMatch.index!;
    const lastLineLength = lastMatch.length;
    const lastLine = text.substr(lastLineOffset + lastLineLength);

    return {
        line: position.start.line + matches.length,
        column: lastLine.length + 1
    };
}

function getParentPosition(node: Parent): Position | undefined {
    let nodeStartPoint: Point | undefined;
    let nodeEndPoint: Point | undefined;

    node.children.forEach((child, index) => {
        if (child.position == null) {
            return;
        }

        nodeEndPoint = child.position.end;

        if (index === 0) {
            // First child
            nodeStartPoint = child.position.start;
        }
    });

    if (nodeStartPoint != null && nodeEndPoint != null) {
        return {
            start: nodeStartPoint,
            end: nodeEndPoint
        };
    }
}

export const relativePositions: Attacher = _options => {
    const transformer: Transformer = tree => {
        let lastPosition: Position = {
            start: {
                line: 1,
                column: 1
            },
            end: {
                line: 1,
                column: 1
            }
        };

        modifyTreeFromDown(tree, node => {
            if (isNodeText(node)) {
                const endPosition: Point = getEndPoint(node, lastPosition);
                const resolvedPosition = {
                    start: {
                        line: lastPosition.start.line,
                        column: lastPosition.start.column
                    },
                    end: endPosition
                };

                lastPosition = {
                    start: endPosition,
                    end: endPosition
                };

                return {
                    ...node,
                    position: resolvedPosition
                };
            } else if (isNodeElement(node) || isNodeRoot(node)) {
                return {
                    ...node,
                    position: getParentPosition(node)
                };
            }

            console.warn(`Not supported Node type '${node.type}'.`);
            return node;
        });

        return tree;
    };

    return transformer;
};
