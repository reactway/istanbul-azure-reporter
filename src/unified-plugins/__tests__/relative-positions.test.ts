import fs from "fs";
import path from "path";
import low from "lowlight";
import unified from "unified";
import os from "os";
import visit from "unist-util-visit";
import { Position } from "unist";

import { relativePositions } from "../relative-positions";

const exampleCode = fs.readFileSync(path.resolve(__dirname, "./example-code.txt"), "utf8");
const exampleCodeLines = exampleCode.split(os.EOL);

function getText(lines: string[], _positon: Position): string {
    const position: Position = {
        start: {
            line: _positon.start.line - 1,
            column: _positon.start.column - 1
        },
        end: {
            line: _positon.end.line - 1,
            column: _positon.end.column - 1
        }
    };
    const startingIndex = position.start.line;
    let result: string[] = [];

    for (let lineIndex = startingIndex; lineIndex <= position.end.line; lineIndex++) {
        const line = lines[lineIndex];

        // Comment symbols explanation:
        // Line            []
        // Text Range      +++
        // Range start/end |
        // Current line    <--
        if (lineIndex === position.start.line && lineIndex === position.end.line) {
            // [ |++++| ] <--
            result.push(line.substring(position.start.column, position.end.column));
        } else if (lineIndex === position.start.line && lineIndex < position.end.line) {
            // [  |+++++] <--
            // [+++++|  ]
            result.push(line.substring(position.start.column, line.length));
        } else if (lineIndex > position.start.line && lineIndex < position.end.line) {
            // [  |+++++]
            // [++++++++] <--
            // [+++++|  ]
            result.push(line);
        } else if (lineIndex > position.start.line && lineIndex === position.end.line) {
            // [  |+++++]
            // [+++++|  ] <--
            result.push(line.substring(0, position.end.column));
        }
    }

    return result.join(os.EOL);
}

it("check positions with source", () => {
    const ast = low.highlight("js", exampleCode).value;

    const astUpdate = unified()
        .use(relativePositions)
        .runSync({ type: "root", children: ast });

    expect(astUpdate).toMatchSnapshot();

    visit(astUpdate, "text", node => {
        const position = node.position!;
        expect(position).toBeDefined();

        const valueFromSource = getText(exampleCodeLines, position);
        expect(node.value).toBe(valueFromSource);
    });
});
