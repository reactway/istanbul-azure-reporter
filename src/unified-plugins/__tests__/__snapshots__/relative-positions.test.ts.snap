// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check positions with source 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 7,
              "line": 1,
            },
            "start": Object {
              "column": 1,
              "line": 1,
            },
          },
          "type": "text",
          "value": "import",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 7,
          "line": 1,
        },
        "start": Object {
          "column": 1,
          "line": 1,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 26,
          "line": 1,
        },
        "start": Object {
          "column": 7,
          "line": 1,
        },
      },
      "type": "text",
      "value": " { createContext } ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 30,
              "line": 1,
            },
            "start": Object {
              "column": 26,
              "line": 1,
            },
          },
          "type": "text",
          "value": "from",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 30,
          "line": 1,
        },
        "start": Object {
          "column": 26,
          "line": 1,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 31,
          "line": 1,
        },
        "start": Object {
          "column": 30,
          "line": 1,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 38,
              "line": 1,
            },
            "start": Object {
              "column": 31,
              "line": 1,
            },
          },
          "type": "text",
          "value": "\\"react\\"",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 38,
          "line": 1,
        },
        "start": Object {
          "column": 31,
          "line": 1,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-string",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 3,
        },
        "start": Object {
          "column": 38,
          "line": 1,
        },
      },
      "type": "text",
      "value": ";

",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 7,
              "line": 3,
            },
            "start": Object {
              "column": 1,
              "line": 3,
            },
          },
          "type": "text",
          "value": "export",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 7,
          "line": 3,
        },
        "start": Object {
          "column": 1,
          "line": 3,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 5,
          "line": 4,
        },
        "start": Object {
          "column": 7,
          "line": 3,
        },
      },
      "type": "text",
      "value": " interface DropdownContextDto {
    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 11,
              "line": 4,
            },
            "start": Object {
              "column": 5,
              "line": 4,
            },
          },
          "type": "text",
          "value": "isOpen",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 11,
          "line": 4,
        },
        "start": Object {
          "column": 5,
          "line": 4,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-attr",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 20,
          "line": 6,
        },
        "start": Object {
          "column": 11,
          "line": 4,
        },
      },
      "type": "text",
      "value": ": boolean;
    isDisabled: boolean;
    onHeaderClick: ",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 22,
                  "line": 6,
                },
                "start": Object {
                  "column": 20,
                  "line": 6,
                },
              },
              "type": "text",
              "value": "()",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 22,
              "line": 6,
            },
            "start": Object {
              "column": 20,
              "line": 6,
            },
          },
          "properties": Object {
            "className": Array [
              "hljs-params",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 25,
              "line": 6,
            },
            "start": Object {
              "column": 22,
              "line": 6,
            },
          },
          "type": "text",
          "value": " =>",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 25,
          "line": 6,
        },
        "start": Object {
          "column": 20,
          "line": 6,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-function",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 26,
          "line": 6,
        },
        "start": Object {
          "column": 25,
          "line": 6,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 30,
              "line": 6,
            },
            "start": Object {
              "column": 26,
              "line": 6,
            },
          },
          "type": "text",
          "value": "void",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 30,
          "line": 6,
        },
        "start": Object {
          "column": 26,
          "line": 6,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 21,
          "line": 7,
        },
        "start": Object {
          "column": 30,
          "line": 6,
        },
      },
      "type": "text",
      "value": ";
    onSectionClick: ",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 23,
                  "line": 7,
                },
                "start": Object {
                  "column": 21,
                  "line": 7,
                },
              },
              "type": "text",
              "value": "()",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 23,
              "line": 7,
            },
            "start": Object {
              "column": 21,
              "line": 7,
            },
          },
          "properties": Object {
            "className": Array [
              "hljs-params",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 26,
              "line": 7,
            },
            "start": Object {
              "column": 23,
              "line": 7,
            },
          },
          "type": "text",
          "value": " =>",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 26,
          "line": 7,
        },
        "start": Object {
          "column": 21,
          "line": 7,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-function",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 27,
          "line": 7,
        },
        "start": Object {
          "column": 26,
          "line": 7,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 31,
              "line": 7,
            },
            "start": Object {
              "column": 27,
              "line": 7,
            },
          },
          "type": "text",
          "value": "void",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 31,
          "line": 7,
        },
        "start": Object {
          "column": 27,
          "line": 7,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 10,
        },
        "start": Object {
          "column": 31,
          "line": 7,
        },
      },
      "type": "text",
      "value": ";
}

",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 7,
              "line": 10,
            },
            "start": Object {
              "column": 1,
              "line": 10,
            },
          },
          "type": "text",
          "value": "export",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 7,
          "line": 10,
        },
        "start": Object {
          "column": 1,
          "line": 10,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 8,
          "line": 10,
        },
        "start": Object {
          "column": 7,
          "line": 10,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 13,
              "line": 10,
            },
            "start": Object {
              "column": 8,
              "line": 10,
            },
          },
          "type": "text",
          "value": "const",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 13,
          "line": 10,
        },
        "start": Object {
          "column": 8,
          "line": 10,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 5,
          "line": 11,
        },
        "start": Object {
          "column": 13,
          "line": 10,
        },
      },
      "type": "text",
      "value": " DropdownContext = createContext<DropdownContextDto>({
    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 11,
              "line": 11,
            },
            "start": Object {
              "column": 5,
              "line": 11,
            },
          },
          "type": "text",
          "value": "isOpen",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 11,
          "line": 11,
        },
        "start": Object {
          "column": 5,
          "line": 11,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-attr",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 13,
          "line": 11,
        },
        "start": Object {
          "column": 11,
          "line": 11,
        },
      },
      "type": "text",
      "value": ": ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 18,
              "line": 11,
            },
            "start": Object {
              "column": 13,
              "line": 11,
            },
          },
          "type": "text",
          "value": "false",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 18,
          "line": 11,
        },
        "start": Object {
          "column": 13,
          "line": 11,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-literal",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 5,
          "line": 12,
        },
        "start": Object {
          "column": 18,
          "line": 11,
        },
      },
      "type": "text",
      "value": ",
    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 15,
              "line": 12,
            },
            "start": Object {
              "column": 5,
              "line": 12,
            },
          },
          "type": "text",
          "value": "isDisabled",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 15,
          "line": 12,
        },
        "start": Object {
          "column": 5,
          "line": 12,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-attr",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 17,
          "line": 12,
        },
        "start": Object {
          "column": 15,
          "line": 12,
        },
      },
      "type": "text",
      "value": ": ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 22,
              "line": 12,
            },
            "start": Object {
              "column": 17,
              "line": 12,
            },
          },
          "type": "text",
          "value": "false",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 22,
          "line": 12,
        },
        "start": Object {
          "column": 17,
          "line": 12,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-literal",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 5,
          "line": 13,
        },
        "start": Object {
          "column": 22,
          "line": 12,
        },
      },
      "type": "text",
      "value": ",
    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 18,
              "line": 13,
            },
            "start": Object {
              "column": 5,
              "line": 13,
            },
          },
          "type": "text",
          "value": "onHeaderClick",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 18,
          "line": 13,
        },
        "start": Object {
          "column": 5,
          "line": 13,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-attr",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 20,
          "line": 13,
        },
        "start": Object {
          "column": 18,
          "line": 13,
        },
      },
      "type": "text",
      "value": ": ",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 22,
                  "line": 13,
                },
                "start": Object {
                  "column": 20,
                  "line": 13,
                },
              },
              "type": "text",
              "value": "()",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 22,
              "line": 13,
            },
            "start": Object {
              "column": 20,
              "line": 13,
            },
          },
          "properties": Object {
            "className": Array [
              "hljs-params",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 25,
              "line": 13,
            },
            "start": Object {
              "column": 22,
              "line": 13,
            },
          },
          "type": "text",
          "value": " =>",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 25,
          "line": 13,
        },
        "start": Object {
          "column": 20,
          "line": 13,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-function",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 26,
          "line": 13,
        },
        "start": Object {
          "column": 25,
          "line": 13,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 35,
              "line": 13,
            },
            "start": Object {
              "column": 26,
              "line": 13,
            },
          },
          "type": "text",
          "value": "undefined",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 35,
          "line": 13,
        },
        "start": Object {
          "column": 26,
          "line": 13,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-literal",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 5,
          "line": 14,
        },
        "start": Object {
          "column": 35,
          "line": 13,
        },
      },
      "type": "text",
      "value": ",
    ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 19,
              "line": 14,
            },
            "start": Object {
              "column": 5,
              "line": 14,
            },
          },
          "type": "text",
          "value": "onSectionClick",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 19,
          "line": 14,
        },
        "start": Object {
          "column": 5,
          "line": 14,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-attr",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 21,
          "line": 14,
        },
        "start": Object {
          "column": 19,
          "line": 14,
        },
      },
      "type": "text",
      "value": ": ",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Object {
                "end": Object {
                  "column": 23,
                  "line": 14,
                },
                "start": Object {
                  "column": 21,
                  "line": 14,
                },
              },
              "type": "text",
              "value": "()",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 23,
              "line": 14,
            },
            "start": Object {
              "column": 21,
              "line": 14,
            },
          },
          "properties": Object {
            "className": Array [
              "hljs-params",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
        Object {
          "position": Object {
            "end": Object {
              "column": 26,
              "line": 14,
            },
            "start": Object {
              "column": 23,
              "line": 14,
            },
          },
          "type": "text",
          "value": " =>",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 26,
          "line": 14,
        },
        "start": Object {
          "column": 21,
          "line": 14,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-function",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 27,
          "line": 14,
        },
        "start": Object {
          "column": 26,
          "line": 14,
        },
      },
      "type": "text",
      "value": " ",
    },
    Object {
      "children": Array [
        Object {
          "position": Object {
            "end": Object {
              "column": 36,
              "line": 14,
            },
            "start": Object {
              "column": 27,
              "line": 14,
            },
          },
          "type": "text",
          "value": "undefined",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 36,
          "line": 14,
        },
        "start": Object {
          "column": 27,
          "line": 14,
        },
      },
      "properties": Object {
        "className": Array [
          "hljs-literal",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    Object {
      "position": Object {
        "end": Object {
          "column": 1,
          "line": 16,
        },
        "start": Object {
          "column": 36,
          "line": 14,
        },
      },
      "type": "text",
      "value": "
});
",
    },
  ],
  "type": "root",
}
`;
